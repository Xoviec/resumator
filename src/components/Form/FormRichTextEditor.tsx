import { useEffect, VoidFunctionComponent } from "react";
import { Controller, useFormContext } from "react-hook-form";
import {
  FormControl,
  FormControlProps,
  InputLabel,
  OutlinedInput,
} from "@mui/material";
import { FormRichTextDraftEditor } from "./FormRichTextDraftEditor";

interface FormRichTextEditorProps extends FormControlProps {
  name: string;
  label: string;
  rows?: number;
}

export const FormRichTextEditor: VoidFunctionComponent<FormRichTextEditorProps> = ({
  name,
  label,
  rows,
  ...props
}) => {
  const { control, resetField } = useFormContext();

  useEffect(() => () => resetField(name), [resetField, name]);

  return (
    <Controller
      defaultValue=""
      control={control}
      name={name!}
      render={({ field: { onChange, value } }) => (
        <FormControl fullWidth size="small" {...props}>
          <InputLabel
            shrink
            variant="outlined"
            htmlFor="form-rich-text-editor-input"
          >
            {label}
          </InputLabel>
          <OutlinedInput
            fullWidth
            notched
            id="form-rich-text-editor-input"
            label={label}
            inputComponent={() => (
              <FormRichTextDraftEditor value={value} onChange={onChange} />
            )}
          />
        </FormControl>
      )}
    />
  );
};
