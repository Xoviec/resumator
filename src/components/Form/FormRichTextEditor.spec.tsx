import { fireEvent, render } from "@testing-library/react";
import { FunctionComponent, useEffect } from "react";
import { FormProvider, useForm } from "react-hook-form";
import { FormRichTextEditor } from "./FormRichTextEditor";

jest.mock("./FormRichTextDraftEditor", () => ({
  ...jest.requireActual("./FormRichTextDraftEditor"),
  FormRichTextDraftEditor: jest
    .requireActual("react")
    .forwardRef((props: any, ref: any) => {
      return (
        <input
          ref={ref}
          data-testid="test-input"
          value={props.value}
          onChange={props.onChange}
        />
      );
    }),
}));

describe("FormRichTextEditor", () => {
  it("should render correctly", () => {
    const FormProviderWrapper: FunctionComponent = ({ children }) => {
      const form = useForm();

      return <FormProvider {...form}>{children}</FormProvider>;
    };

    const { asFragment } = render(
      <FormProviderWrapper>
        <FormRichTextEditor name="test-form-name" label="test-form-label" />
      </FormProviderWrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should update field when the value changes", () => {
    const FormProviderWrapper: FunctionComponent<any> = ({ children, onChange }) => {
      const form = useForm();
      const fieldValue = form.watch("test-form-name");

      useEffect(() => {
        onChange(fieldValue);
      }, [fieldValue, onChange]);

      return <FormProvider {...form}>{children}</FormProvider>;
    };
    const mockChange = jest.fn();

    const { getByTestId } = render(
      <FormProviderWrapper onChange={mockChange}>
        <FormRichTextEditor name="test-form-name" label="test-form-label" />
      </FormProviderWrapper>
    );

    fireEvent.change(getByTestId("test-input"), {
      target: { value: "test-input-value" },
    });

    expect(mockChange).toHaveBeenCalledWith("test-input-value");
  });
});
