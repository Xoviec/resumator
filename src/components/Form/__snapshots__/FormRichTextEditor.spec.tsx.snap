// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormRichTextEditor should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for="form-rich-text-editor-input"
    >
      test-form-label
    </label>
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        data-testid="test-input"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-1in441m"
        >
          <span>
            test-form-label
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
